// @graphenedata/cli/dist/ui/internal/pageInputs.svelte.js
import { createContext } from "svelte";
class BoundField {
  constructor(controller, name, codec) {
    this.controller = controller;
    this.name = name;
    this.codec = codec;
    this.value = $state(codec.empty());
    this.hasExternalValue = $state(false);
    this.controller.registerField(this);
    this.sync(this.controller.getParams(), "init");
  }
  value;
  hasExternalValue;
  keys() {
    return this.codec.keys(this.name);
  }
  sync(params, source) {
    let next = this.codec.read(params, this.name);
    this.value = next.value;
    if (source === "init") this.hasExternalValue = next.present;
    else if (source === "external") this.hasExternalValue = true;
  }
  normalize(params) {
    this.codec.normalize?.(params, this.name);
  }
  set(next) {
    let nextParams = this.codec.write(this.name, next);
    if (typeof window !== "undefined" && window.$GRAPHENE?.updateParams) window.$GRAPHENE.updateParams(nextParams);
    else this.controller.updateParams(nextParams);
  }
  destroy() {
    this.controller.unregisterField(this);
  }
}
const textCodec = {
  keys: (name) => [name],
  empty: () => "",
  read(params, name) {
    let raw = params[name];
    if (raw === void 0 || raw === null) return { present: false, value: "" };
    if (Array.isArray(raw)) return { present: raw.length > 0, value: raw.length ? String(raw[0] ?? "") : "" };
    return { present: true, value: String(raw) };
  },
  write(name, value) {
    return { [name]: value === "" ? null : value };
  }
};
const dropdownSingleCodec = {
  keys: (name) => [name],
  empty: () => [],
  read(params, name) {
    let raw = params[name];
    if (raw === void 0 || raw === null) return { present: false, value: [] };
    if (Array.isArray(raw)) return { present: raw.length > 0, value: raw.length ? [raw[0]] : [] };
    return { present: true, value: [raw] };
  },
  write(name, value) {
    return { [name]: value.length ? value[0] : null };
  }
};
const dropdownMultiCodec = {
  keys: (name) => [name],
  empty: () => [],
  read(params, name) {
    let raw = params[name];
    if (raw === void 0 || raw === null) return { present: false, value: [] };
    if (Array.isArray(raw)) return { present: raw.length > 0, value: [...raw] };
    return { present: true, value: [raw] };
  },
  write(name, value) {
    return { [name]: value.length ? [...value] : null };
  },
  normalize(params, name) {
    let raw = params[name];
    if (raw === void 0 || raw === null || Array.isArray(raw)) return;
    params[name] = [raw];
  }
};
const dateRangeCodec = {
  keys: (name) => [`${name}_start`, `${name}_end`],
  empty: () => ({ start: null, end: null }),
  read(params, name) {
    let startRaw = params[`${name}_start`];
    let endRaw = params[`${name}_end`];
    let start = readScalar(startRaw);
    let end = readScalar(endRaw);
    return { present: start !== null || end !== null, value: { start, end } };
  },
  write(name, value) {
    return {
      [`${name}_start`]: value.start,
      [`${name}_end`]: value.end
    };
  }
};
function readScalar(value) {
  if (value === void 0 || value === null) return null;
  if (Array.isArray(value)) return value.length ? String(value[0] ?? "") : null;
  return String(value);
}
function cloneParamValue(value) {
  if (Array.isArray(value)) return [...value];
  return value;
}
function cloneParams(params) {
  return Object.fromEntries(Object.entries(params).map(([name, value]) => [name, cloneParamValue(value)]));
}
function paramValueEqual(left, right) {
  if (Array.isArray(left) || Array.isArray(right)) {
    if (!Array.isArray(left) || !Array.isArray(right)) return false;
    if (left.length !== right.length) return false;
    return left.every((value, index) => value === right[index]);
  }
  return left === right;
}
function getChangedKeys(left, right) {
  let changed = /* @__PURE__ */ new Set();
  let keys = /* @__PURE__ */ new Set([...Object.keys(left), ...Object.keys(right)]);
  keys.forEach((key) => {
    if (!paramValueEqual(left[key], right[key])) changed.add(key);
  });
  return changed;
}
class PageInputs {
  params;
  subscribers;
  fieldsByKey;
  onPopState;
  constructor() {
    this.params = $state.raw({});
    this.subscribers = /* @__PURE__ */ new Set();
    this.fieldsByKey = /* @__PURE__ */ new Map();
    this.params = cloneParams(this.readFromUrl());
    if (typeof window !== "undefined") {
      this.onPopState = () => this.syncFromUrl();
      window.addEventListener("popstate", this.onPopState);
    }
  }
  destroy() {
    if (this.onPopState) window.removeEventListener("popstate", this.onPopState);
    this.subscribers.clear();
    this.fieldsByKey.clear();
  }
  registerField(field) {
    field.keys().forEach((key) => {
      let fields = this.fieldsByKey.get(key);
      if (!fields) {
        fields = /* @__PURE__ */ new Set();
        this.fieldsByKey.set(key, fields);
      }
      fields.add(field);
    });
  }
  unregisterField(field) {
    field.keys().forEach((key) => {
      let fields = this.fieldsByKey.get(key);
      if (!fields) return;
      fields.delete(field);
      if (fields.size === 0) this.fieldsByKey.delete(key);
    });
  }
  text(name) {
    return new BoundField(this, name, textCodec);
  }
  dropdown(name, multiple) {
    return new BoundField(this, name, multiple ? dropdownMultiCodec : dropdownSingleCodec);
  }
  dateRange(name) {
    return new BoundField(this, name, dateRangeCodec);
  }
  getParam(name) {
    return this.params[name];
  }
  getParams() {
    return cloneParams(this.params);
  }
  subscribeParams(subscriber) {
    this.subscribers.add(subscriber);
    return () => this.subscribers.delete(subscriber);
  }
  updateParam(name, value) {
    this.updateParams({ [name]: value });
  }
  updateParams(nextParams) {
    let merged = { ...this.params };
    Object.entries(nextParams).forEach(([name, value]) => {
      merged[name] = Array.isArray(value) ? [...value] : value;
    });
    this.applySnapshot(merged, "local");
  }
  reset() {
    let changed = new Set(Object.keys(this.params));
    if (changed.size === 0) return;
    this.params = {};
    this.syncUrl();
    this.syncFields(changed, "local");
    let snapshot = this.getParams();
    this.subscribers.forEach((subscriber) => subscriber(snapshot, { changed }));
  }
  syncFromUrl() {
    this.applySnapshot(this.readFromUrl(), "external");
  }
  applySnapshot(nextParams, source) {
    let cloned = cloneParams(nextParams);
    this.normalizeSnapshot(cloned);
    let changed = getChangedKeys(this.params, cloned);
    if (changed.size === 0) return;
    this.params = cloned;
    if (source !== "external") this.syncUrl();
    this.syncFields(changed, source);
    let snapshot = this.getParams();
    this.subscribers.forEach((subscriber) => subscriber(snapshot, { changed }));
    window.$GRAPHENE?.rerunQueries?.();
  }
  syncFields(changed, source) {
    let fields = /* @__PURE__ */ new Set();
    changed.forEach((key) => {
      this.fieldsByKey.get(key)?.forEach((field) => fields.add(field));
    });
    fields.forEach((field) => field.sync(this.params, source));
  }
  normalizeSnapshot(params) {
    let fields = /* @__PURE__ */ new Set();
    this.fieldsByKey.forEach((keyFields) => {
      keyFields.forEach((field) => fields.add(field));
    });
    fields.forEach((field) => field.normalize(params));
  }
  syncUrl() {
    if (typeof window === "undefined") return;
    let search = new URLSearchParams();
    Object.entries(this.params).forEach(([name, value]) => {
      if (Array.isArray(value)) {
        value.forEach((item) => search.append(name, String(item)));
        return;
      }
      if (value === null || value === void 0 || value === "") return;
      search.append(name, String(value));
    });
    let nextSearch = search.toString();
    let currentSearch = window.location.search.replace(/^\?/, "");
    if (nextSearch === currentSearch) return;
    let nextUrl = window.location.pathname + (nextSearch ? `?${nextSearch}` : "") + window.location.hash;
    window.history.replaceState(window.history.state, "", nextUrl);
  }
  readFromUrl() {
    if (typeof window === "undefined") return {};
    let next = {};
    for (let [name, value] of new URLSearchParams(window.location.search).entries()) {
      let existing = next[name];
      if (existing === void 0) next[name] = value;
      else if (Array.isArray(existing)) existing.push(value);
      else next[name] = [String(existing), value];
    }
    return next;
  }
}
const [getPageInputsContext, setPageInputsContext] = createContext();
let activePageInputs = null;
function activatePageInputs(pageInputs) {
  if (activePageInputs && activePageInputs !== pageInputs) activePageInputs.destroy();
  activePageInputs = pageInputs;
  return pageInputs;
}
function getActivePageInputs() {
  if (!activePageInputs) activePageInputs = new PageInputs();
  return activePageInputs;
}
function releasePageInputs(pageInputs) {
  if (activePageInputs !== pageInputs) return;
  pageInputs.destroy();
  activePageInputs = null;
}
function getPageInputs() {
  try {
    return getPageInputsContext();
  } catch {
    return getActivePageInputs();
  }
}
function captureInitial(factory) {
  return factory();
}
export {
  PageInputs,
  activatePageInputs,
  captureInitial,
  getActivePageInputs,
  getPageInputs,
  releasePageInputs,
  setPageInputsContext
};
