// @graphenedata/cli/dist/cli/bigQuery-YIWXZPY6.js
import {
  config
} from "./chunk-SQVXTHE5.js";

// connections/bigQuery.ts
import { BigQuery, BigQueryDate, BigQueryTimestamp } from "@google-cloud/bigquery";
function validateBigQueryIdent(ident) {
  if (!/^[\w.-]+$/.test(ident)) throw new Error(`Invalid BigQuery identifier: ${ident}`);
}
var BigQueryConnection = class {
  client;
  projectId;
  defaultNamespace;
  constructor(options = {}) {
    options.projectId ||= config.bigquery?.projectId;
    if (!options.projectId) throw new Error("projectId must be set in config or provided in service account credentials");
    this.projectId = options.projectId;
    this.client = new BigQuery({ ...options, userAgent: "Graphene" });
    this.defaultNamespace = config.defaultNamespace;
  }
  async runQuery(sql, params) {
    let [job] = await this.client.createQueryJob({ query: sql, useLegacySql: false, params });
    let [rows] = await job.getQueryResults({ maxResults: 1e4 });
    let metadata = job.metadata || (await job.getMetadata())[0];
    let totalRows = Number(metadata?.statistics?.query?.totalRows ?? rows.length);
    rows.forEach((r) => {
      Object.entries(r).forEach(([k, v]) => {
        if (v instanceof BigQueryTimestamp) r[k] = v.value;
        if (v instanceof BigQueryDate) r[k] = v.value;
      });
    });
    return { rows, totalRows };
  }
  async listDatasets() {
    let [datasets] = await this.client.getDatasets();
    return datasets.map((d) => String(d.id || d.metadata.datasetReference?.datasetId || "").toLowerCase());
  }
  async listTables(dataset) {
    if (!dataset) throw new Error("BigQuery requires a dataset");
    validateBigQueryIdent(dataset);
    let resolvedDataset = await this.resolveDatasetName(dataset);
    let res = await this.runQuery(`select table_name as table_name
      from \`${resolvedDataset}.INFORMATION_SCHEMA.TABLES\`
      where table_type in ('BASE TABLE', 'VIEW') order by table_name`);
    return res.rows.map((r) => `${resolvedDataset.toLowerCase()}.${String(r["table_name"]).toLowerCase()}`);
  }
  async describeTable(target) {
    let parts = target.split(".");
    let table = parts.pop() || "";
    let dataset = parts.join(".") || this.defaultNamespace;
    if (!dataset) throw new Error("No dataset specified and no default namespace configured");
    validateBigQueryIdent(dataset);
    let resolvedDataset = await this.resolveDatasetName(dataset);
    let sql = `
      select column_name as column_name, data_type as data_type, ordinal_position as ordinal_position
      from \`${resolvedDataset}.INFORMATION_SCHEMA.COLUMNS\`
      where lower(table_name) = lower(@table)
      order by ordinal_position
    `.trim();
    let res = await this.runQuery(sql, { table });
    return res.rows.map((row) => {
      return { name: String(row["column_name"]).toLowerCase(), dataType: String(row["data_type"]) };
    });
  }
  async resolveDatasetName(name) {
    let datasets = await this.listDatasets();
    return datasets.find((ds) => ds.toLowerCase() == name.toLowerCase()) || name;
  }
  async close() {
  }
};
export {
  BigQueryConnection
};
