// @graphenedata/cli/dist/cli/chunk-SQVXTHE5.js
// ../lang/config.ts
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import path from "path";
var config = { dialect: "duckdb", root: "" };
function setGlobalConfig(cfg) {
  Object.keys(config).forEach((key) => delete config[key]);
  Object.assign(config, normalizeConfig(cfg));
}
function normalizeConfig(input, defaultRoot = process.cwd()) {
  let cfg = { ...input };
  if (cfg.namespace && !cfg.defaultNamespace) cfg.defaultNamespace = cfg.namespace;
  let dialect = cfg.dialect || "duckdb";
  if (cfg.bigquery) dialect = "bigquery";
  else if (cfg.snowflake) dialect = "snowflake";
  else if (cfg.clickhouse) dialect = "clickhouse";
  else if (cfg.duckdb) dialect = "duckdb";
  let envFile = [".env"];
  if (Array.isArray(cfg.envFile)) envFile = cfg.envFile;
  else if (cfg.envFile) envFile = [cfg.envFile];
  return {
    ...cfg,
    dialect,
    root: path.resolve(cfg.root || defaultRoot),
    port: cfg.port || Number(process.env.GRAPHENE_PORT) || 4e3,
    ignoredFiles: cfg.ignoredFiles || [],
    envFile
  };
}
async function loadConfig(dir, envLoader) {
  let configDir = path.resolve(dir);
  while (!existsSync(path.join(configDir, "package.json"))) {
    let parent = path.dirname(configDir);
    if (parent == configDir) throw new Error(`No package.json found in ${path.resolve(dir)} or its parents`);
    configDir = parent;
  }
  let txt = await readFile(path.join(configDir, "package.json"), "utf8");
  let graphene = JSON.parse(txt).graphene;
  if (!graphene || typeof graphene != "object" || Array.isArray(graphene)) {
    throw new Error(`No graphene config found in ${path.join(configDir, "package.json")}`);
  }
  let envFiles = Array.isArray(graphene.envFile) ? graphene.envFile : [graphene.envFile || ".env"];
  envLoader(envFiles.map((file) => path.resolve(configDir, file)));
  let cfg = normalizeConfig({ ...graphene, root: configDir }, configDir);
  return cfg;
}

export {
  config,
  setGlobalConfig,
  loadConfig
};
