// @graphenedata/cli/dist/cli/clickhouse-S3BJSKND.js
// connections/clickhouse.ts
import { createClient } from "@clickhouse/client";
var ClickHouseConnection = class {
  client;
  defaultDatabase;
  constructor(options) {
    this.defaultDatabase = options.database || "default";
    this.client = createClient({
      url: options.url,
      username: options.username,
      password: options.password,
      database: this.defaultDatabase,
      application: "Graphene",
      request_timeout: options.requestTimeout
    });
  }
  async runQuery(sql, _params) {
    let result = await this.client.query({ query: sql, format: "JSONEachRow" });
    let rows = await result.json();
    return { rows, totalRows: rows.length };
  }
  async listDatasets() {
    let res = await this.runQuery(`
      select name
      from system.databases
      where lower(name) not in ('system', 'information_schema')
      order by name
    `);
    return res.rows.map((row) => String(row["name"]).toLowerCase());
  }
  async listTables(database = this.defaultDatabase) {
    let sql = `
      select database, name
      from system.tables
      where lower(database) = lower('${escapeClickHouseString(database)}')
      order by name
    `.trim();
    let res = await this.runQuery(sql);
    return res.rows.map((row) => `${String(row["database"]).toLowerCase()}.${String(row["name"]).toLowerCase()}`);
  }
  async describeTable(target) {
    let parts = target.split(".").filter(Boolean);
    let table = parts.pop() || "";
    let database = parts.join(".") || this.defaultDatabase;
    let sql = `
      select name, type, position
      from system.columns
      where lower(database) = lower('${escapeClickHouseString(database)}')
        and lower(table) = lower('${escapeClickHouseString(table)}')
      order by position
    `.trim();
    let res = await this.runQuery(sql);
    return res.rows.map((row) => ({ name: String(row["name"]).toLowerCase(), dataType: String(row["type"]) }));
  }
  async close() {
    await this.client.close();
  }
};
function escapeClickHouseString(value) {
  return value.replace(/\\/g, "\\\\").replace(/'/g, "\\'");
}
export {
  ClickHouseConnection
};
