// @graphenedata/cli/dist/cli/duckdb-V6PJEA7H.js
import {
  config
} from "./chunk-SQVXTHE5.js";

// connections/duckdb.ts
import { DuckDBTimestampValue, DuckDBInstance, DuckDBDateValue, DuckDBDecimalValue } from "@duckdb/node-api";
import { promises as fs } from "fs";
import path from "path";
var DuckDBConnection = class {
  options;
  ready;
  connection = null;
  constructor(options) {
    this.options = options || {};
    this.ready = this.initialize();
  }
  async initialize() {
    let dbPath = this.options.path || config.duckdb?.path;
    if (!dbPath) {
      let files = await fs.readdir(config.root);
      dbPath = files.find((f) => f.endsWith(".duckdb"));
      if (!dbPath) throw new Error("No .duckdb file found in current directory");
    }
    if (!path.isAbsolute(dbPath)) dbPath = path.resolve(config.root, dbPath);
    let db = await DuckDBInstance.create(":memory:");
    this.connection = await db.connect();
    let escapedPath = dbPath.replace(/'/g, "''");
    await this.connection.run(`attach '${escapedPath}' as graphene_cli (READ_ONLY);`);
    await this.connection.run("use graphene_cli;");
  }
  async runQuery(sql, params) {
    await this.ready;
    let reader = params ? await this.connection.runAndReadAll(sql, params) : await this.connection.runAndReadAll(sql);
    let rows = reader.getRowObjects().map((record) => {
      let out = {};
      for (let [k, v] of Object.entries(record)) {
        if (typeof v === "bigint") out[k] = Number(v);
        else if (v === null) out[k] = null;
        else if (v instanceof DuckDBTimestampValue) out[k] = new Date(Number(v.micros / 1000n)).toUTCString();
        else if (v instanceof DuckDBDateValue) out[k] = v.toString();
        else if (v instanceof DuckDBDecimalValue) out[k] = v.toDouble();
        else if (typeof v === "object") throw new Error(`Unsupported datatype ${v.constructor?.name}`);
        else out[k] = v;
      }
      return out;
    });
    return { rows };
  }
  async listDatasets() {
    return await Promise.resolve([]);
  }
  async listTables() {
    let sql = `
      select table_schema as table_schema, table_name as table_name
      from information_schema.tables
      where table_type in ('BASE TABLE', 'VIEW') and table_schema not in ('information_schema', 'pg_catalog')
      order by table_schema, table_name
    `.trim();
    let res = await this.runQuery(sql);
    return res.rows.map((row) => String(row["table_name"]).toLowerCase());
  }
  async describeTable(target) {
    let parts = target.split(".");
    let table = parts.pop() || "";
    let schema = parts[0];
    let schemaFilter = schema ? "lower(table_schema) = lower($2)" : "table_schema not in ('information_schema', 'pg_catalog')";
    let sql = `
      select column_name as column_name, data_type as data_type, ordinal_position as ordinal_position
      from information_schema.columns
      where lower(table_name) = lower($1) and ${schemaFilter}
      order by ordinal_position
    `.trim();
    let params = schema ? [table, schema] : [table];
    let res = await this.runQuery(sql, params);
    return res.rows.map((row) => {
      return { name: String(row["column_name"]).toLowerCase(), dataType: String(row["data_type"]) };
    });
  }
  async close() {
    await this.ready;
    this.connection?.closeSync();
  }
};
export {
  DuckDBConnection
};
