// @graphenedata/cli/dist/cli/serve2-CGQSM7TD.js
import {
  analyzeWorkspace,
  getWorkspaceScanCounts,
  loadWorkspace,
  mockFileMap,
  runQuery,
  runVitePlugin,
  toSql
} from "./chunk-UTV3ERGI.js";
import {
  config
} from "./chunk-SQVXTHE5.js";

// serve2.ts
import { svelte, vitePreprocess } from "@sveltejs/vite-plugin-svelte";
import crypto from "crypto";
import fs2 from "fs-extra";
import { mdsvex } from "mdsvex";
import { createRequire } from "module";
import path2 from "path";
import { fileURLToPath } from "url";
import { createServer, optimizeDeps, resolveConfig } from "vite";

// mdCompile.ts
import { decodeHTML } from "entities";
import fs from "fs";
import yaml from "js-yaml";
import JSON5 from "json5";
import path from "path";
import sanitizeHtml from "sanitize-html";
import { visit } from "unist-util-visit";
function svelteStringAttr(str) {
  let literal = str.replace(/\\/g, "\\\\").replace(/`/g, "\\`").replace(/\$\{/g, "\\${").replace(/\r/g, "\\r").replace(/\n/g, "\\n").replace(/&/g, "\\u0026").replace(/"/g, "\\u0022").replace(/</g, "\\u003c").replace(/>/g, "\\u003e");
  return `{\`${literal}\`}`;
}
function liftInlineEChartsConfig(content) {
  return content.replace(/<ECharts\b([^>]*)>([\s\S]*?)<\/ECharts>/g, (match, attrs = "", body = "") => {
    let inline = body.trim();
    if (!inline) return match;
    if (/\sconfig\s*=/.test(attrs)) return match;
    let source = inline.startsWith("{") ? inline : `{${inline}}`;
    let config2 = JSON.stringify(JSON5.parse(source), (_key, value) => typeof value == "string" ? decodeHTML(value) : value);
    return `<ECharts${attrs} config={${config2}}></ECharts>`;
  });
}
function extractQueries() {
  return function transformer(tree) {
    visit(tree, "code", (node, index, parent) => {
      if (index === null) return;
      let name = typeof node.meta === "string" ? node.meta : "";
      let code = typeof node.value === "string" ? node.value.trim() : "";
      parent.children[index] = { type: "html", value: `<GrapheneQuery name="${svelteStringAttr(name)}" code="${svelteStringAttr(code)}" />` };
    });
  };
}
function escapeAngles() {
  return function transformer(tree) {
    visit(tree, "text", (node) => {
      if (!node.value || typeof node.value !== "string") return;
      if (!node.value.includes("<")) return;
      node.value = node.value.replace(/</g, "&lt;");
    });
  };
}
function mergeAdjacentHtml() {
  return function transformer(tree) {
    visit(tree, (parent) => {
      if (!Array.isArray(parent?.children)) return;
      for (let i = 0; i < parent.children.length; i++) {
        if (parent.children[i]?.type !== "html") continue;
        let j = i;
        while (j + 1 < parent.children.length && parent.children[j + 1]?.type === "html") j++;
        if (j == i) continue;
        let value = parent.children.slice(i, j + 1).map((node) => node.value || "").join("\n");
        parent.children.splice(i, j - i + 1, { type: "html", value });
      }
    });
  };
}
function sanitizeMarkdown() {
  return function transformer(tree) {
    visit(tree, "raw", (node) => {
      if (typeof node.value !== "string") return;
      let expanded = node.value.replace(/<(\w+)((?:\s[^<>]*?)?)\s*\/>/gi, (_, name, attrs = "") => {
        let spacing = attrs;
        return `<${name}${spacing}></${name}>`;
      });
      let sanitized = sanitizeHtml(expanded, {
        ...sanitizeHtml.defaults,
        allowedTags: [...sanitizeHtml.defaults.allowedTags, ...componentNames()],
        allowedAttributes: {
          ...sanitizeHtml.defaults.allowedAttributes,
          ...Object.fromEntries(componentNames().map((n) => [n, ["*"]]))
        },
        parser: {
          ...sanitizeHtml.defaults.parser || {},
          lowerCaseAttributeNames: false,
          lowerCaseTags: false
        }
      });
      node.value = sanitized;
    });
  };
}
function injectComponentImports() {
  let imp = `const {${componentNames().join(", ")}} = window.$GRAPHENE.components`;
  return {
    markup: ({ content, filename }) => {
      if (!filename.endsWith(".md")) return;
      content = liftInlineEChartsConfig(content);
      if (content.includes("<script>")) {
        content = content.replace("<script>", `<script>
${imp}`);
      } else {
        content = `<script>
${imp}
</script>
${content}`;
      }
      return { code: content };
    },
    style: () => {
    },
    script: () => {
    }
  };
}
var cachedComponentNames = null;
function componentNames() {
  if (cachedComponentNames) return cachedComponentNames;
  let files = fs.readdirSync(path.join(import.meta.dirname, "../ui/components"));
  cachedComponentNames = files.map((f) => path.basename(f, ".svelte")).filter((f) => !f.startsWith("_"));
  return cachedComponentNames || [];
}
var frontmatterRe = /^---\s*\n([\s\S]*?)\n---(?:\n|$)/;
function extractFrontmatter(contents) {
  let match = contents.trimStart().match(frontmatterRe);
  if (!match) return {};
  let raw = yaml.safeLoad(match[1]);
  return { title: raw?.title ? String(raw.title) : void 0 };
}
var remarkPlugins = [extractQueries, escapeAngles, mergeAdjacentHtml];
var rehypePlugins = [sanitizeMarkdown];

// serve2.ts
var svelteWarnings = [];
function clearSvelteWarnings() {
  svelteWarnings.length = 0;
}
var QUERY_VERSION = 1;
var uiRoot;
var nodeRequire = createRequire(import.meta.url);
async function serve2(telemetry) {
  let server = await createServer(await createConfig(telemetry));
  await server.listen();
  console.log(`Server running at http://localhost:${server.config.server.port}`);
  return server;
}
async function createConfig(telemetry) {
  uiRoot = path2.join(fileURLToPath(import.meta.url), "../../ui");
  let port = Number(process.env.GRAPHENE_PORT) || 4e3;
  let svelteRoot = path2.dirname(nodeRequire.resolve("svelte/package.json"));
  let sveltePackage = nodeRequire("svelte/package.json");
  let svelteDependencyRoot = path2.dirname(svelteRoot);
  let svelteExport = (name) => path2.join(svelteRoot, sveltePackage.exports[name].browser || sveltePackage.exports[name].default);
  let packaged = path2.basename(path2.dirname(uiRoot)) == "dist";
  await fs2.ensureDir(path2.resolve(config.root, "node_modules/.graphene"));
  let inContainer = fs2.existsSync("/.dockerenv");
  let host = inContainer ? "0.0.0.0" : "127.0.0.1";
  return {
    root: config.root,
    logLevel: process.env.NODE_ENV == "test" ? "silent" : "info",
    plugins: [
      svelte({
        configFile: false,
        extensions: [".svelte", ".md"],
        preprocess: [
          vitePreprocess(),
          mdsvex({
            extensions: [".md"],
            remarkPlugins,
            rehypePlugins
          }),
          injectComponentImports()
        ],
        onwarn(warning, defaultHandler) {
          if (process.env.NODE_ENV === "test") {
            svelteWarnings.push({ code: warning.code, message: warning.message, filename: warning.filename });
          }
          defaultHandler?.(warning);
        }
      }),
      fixSvelteDepsInTests(),
      fixHmrForFailedModules(),
      runVitePlugin(),
      handleRequestPlugin,
      updateWorkspacePlugin(telemetry),
      mockFilesForTests()
    ],
    publicDir: path2.resolve(uiRoot, "public"),
    // on the fence about this one. This would make it less likely we need to optimize when alternating between dev and tests.
    // cacheDir: process.env.NODE_ENV == 'test' ? 'node_modules/.vite-tests' : 'node_modules/.vite',
    server: {
      port,
      host,
      fs: { strict: false },
      strictPort: true,
      hmr: { overlay: false }
      // we handle compilation errors ourselves (see LocalApp.svelte)
    },
    resolve: {
      alias: [
        { find: /^graphene$/, replacement: path2.resolve(uiRoot, "web.js") },
        // Vite runs in a user project, but svelte is a direct dependency of the cli, and thus transitive to the user project.
        // So when Vite tries to resolve `svelte` from a compiled md page, it can't find it without these aliases.
        { find: /^svelte$/, replacement: svelteExport(".") },
        { find: /^svelte\/animate$/, replacement: svelteExport("./animate") },
        { find: /^svelte\/attachments$/, replacement: svelteExport("./attachments") },
        { find: /^svelte\/easing$/, replacement: svelteExport("./easing") },
        { find: /^svelte\/events$/, replacement: svelteExport("./events") },
        { find: /^svelte\/internal$/, replacement: svelteExport("./internal") },
        { find: /^svelte\/internal\/client$/, replacement: svelteExport("./internal/client") },
        { find: /^svelte\/internal\/disclose-version$/, replacement: svelteExport("./internal/disclose-version") },
        { find: /^svelte\/internal\/flags\/async$/, replacement: svelteExport("./internal/flags/async") },
        { find: /^svelte\/internal\/flags\/legacy$/, replacement: svelteExport("./internal/flags/legacy") },
        { find: /^svelte\/internal\/flags\/tracing$/, replacement: svelteExport("./internal/flags/tracing") },
        { find: /^svelte\/legacy$/, replacement: svelteExport("./legacy") },
        { find: /^svelte\/motion$/, replacement: svelteExport("./motion") },
        { find: /^svelte\/reactivity$/, replacement: svelteExport("./reactivity") },
        { find: /^svelte\/reactivity\/window$/, replacement: svelteExport("./reactivity/window") },
        { find: /^svelte\/store$/, replacement: svelteExport("./store") },
        { find: /^svelte\/transition$/, replacement: svelteExport("./transition") },
        { find: /^clsx$/, replacement: path2.join(svelteDependencyRoot, "clsx/dist/clsx.mjs") }
      ]
    },
    optimizeDeps: {
      noDiscovery: process.env.NODE_ENV == "test",
      // tests manually optimize before starting test workers
      exclude: ["virtual:nav"],
      // provided by a plugin, so don't try and optimize it
      // Vite running in a user project will not naturally discover and optimize these transitive deps.
      // When you launch the server, your first page load will automatically refresh after a second or two as Vite now sees and optimizes these.
      // This line makes it do that up-front, avoiding that reload jank. The packaged CLI also pre-bundles the `graphene` alias itself;
      // doing that from source causes trouble in examples/tests because the alias points outside node_modules.
      // `graphene` here is a special case: when packaged up it is considered a dependency, but in examples/tests, including it would cause errors.
      // oxfmt-ignore
      include: [
        ...packaged ? ["graphene"] : [],
        "@graphenedata/cli > svelte",
        "@graphenedata/cli > chroma-js",
        "@graphenedata/cli > echarts",
        "@graphenedata/cli > @graphenedata/html2canvas",
        "@graphenedata/cli > @graphenedata/ui > svelte",
        "@graphenedata/cli > @graphenedata/ui > chroma-js",
        "@graphenedata/cli > @graphenedata/ui > echarts/dist/echarts.esm.js",
        "@graphenedata/cli > @graphenedata/ui > @graphenedata/html2canvas"
      ]
    }
  };
}
async function handleQuery(req, res) {
  let chunks = [];
  for await (let chunk of req) chunks.push(chunk);
  let { gsql, params, hashes } = JSON.parse(Buffer.concat(chunks).toString());
  res.setHeader("Content-Type", "application/json");
  await workspaceLoadPromise;
  let gsqlFiles = workspaceFiles.filter((file) => !file.path.endsWith(".md"));
  let result = analyzeWorkspace({ config, files: [...gsqlFiles, { path: "input", contents: gsql }] });
  updateParsedFiles(result);
  let diagnostics = result.diagnostics;
  if (diagnostics.length) {
    res.statusCode = 400;
    res.end(JSON.stringify(diagnostics[0]));
    return;
  }
  let queries = result.files.find((file) => file.path == "input")?.queries || [];
  if (queries.length > 1) throw new Error("Found multiple queries, which could be a parsing error");
  let sql = toSql(queries[0], params);
  let hash = crypto.createHash("SHA1").update(`query-v${QUERY_VERSION}|${sql}`).digest("hex");
  res.setHeader("ETag", hash);
  if (hashes.includes(hash) && req.headers["cache-control"] != "no-cache") {
    res.statusCode = 304;
    return res.end();
  }
  let queryResults = await runQuery(sql);
  let totalRows = queryResults.totalRows ?? queryResults.rows.length;
  if (totalRows > queryResults.rows.length) throw new Error("Query returns too many rows");
  let fields = queries[0].fields.map((field) => ({ name: field.name, type: field.type, metadata: field.metadata || {} }));
  res.end(JSON.stringify({ rows: queryResults.rows, hash, fields, sql }));
}
async function handlePage(server, res) {
  res.setHeader("Content-Type", "text/html");
  let html = await server.transformIndexHtml(
    "/index.html",
    `<!doctype html>
  <html lang="en">
    <head>
      <meta charset="UTF-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>Graphene</title>
      <link rel="icon" href="/favicon.ico" />
    </head>
    <body>
      <script type="module">
        import 'graphene'
      </script>
    </body>
  </html>`
  );
  return res.end(html);
}
async function prepareDeps() {
  let cfg = await resolveConfig(await createConfig(), "serve");
  await optimizeDeps(cfg, true);
}
function fixSvelteDepsInTests() {
  let viteConfig;
  function configResolved(cfg) {
    viteConfig = cfg;
  }
  function buildStart() {
    if (process.env.NODE_ENV != "test") return;
    viteConfig.optimizeDeps.force = false;
  }
  buildStart.sequential = true;
  return { name: "fix-svelte-deps", enforce: "post", configResolved, buildStart };
}
function fixHmrForFailedModules() {
  return {
    name: "fix-hmr-for-failed-modules",
    hotUpdate({ modules }) {
      let hasFailed = modules.some((m) => !m.transformResult);
      if (hasFailed) {
        this.environment.hot.send({ type: "full-reload", path: "*" });
        return [];
      }
    }
  };
}
var workspaceLoadPromise;
var workspaceFiles = [];
var mdFiles = [];
function updateWorkspacePlugin(telemetry) {
  return {
    name: "updateWorkspace",
    resolveId(id) {
      if (id == "virtual:nav") return "\0virtual:nav";
    },
    load(id) {
      if (id != "\0virtual:nav") return;
      let res = [...mdFiles];
      if (process.env.NODE_ENV == "test") {
        for (let [path3, contents] of Object.entries(mockFileMap)) {
          let mockFile = { path: path3, title: extractFrontmatter(contents).title };
          let idx = res.findIndex((file) => file.path == path3);
          if (idx >= 0) res.splice(idx, 1, mockFile);
          else res.push(mockFile);
        }
      }
      return `export default ${JSON.stringify(res)}`;
    },
    configureServer: (s) => {
      let refresh = async () => {
        workspaceLoadPromise = (async () => {
          let loaded = await loadWorkspace(config.root, true, config.ignoredFiles);
          telemetry?.event("workspace_scanned", { command: "serve", ...getWorkspaceScanCounts(loaded) });
          workspaceFiles = loaded.map((file) => {
            let existing = workspaceFiles.find((existing2) => existing2.path == file.path && existing2.contents == file.contents);
            return existing?.parsed ? { ...file, parsed: existing.parsed } : file;
          });
        })();
        await workspaceLoadPromise;
        mdFiles = workspaceFiles.filter((file) => file.path.endsWith(".md")).map((f) => ({ path: f.path, title: extractFrontmatter(f.contents).title }));
        let mod = s.moduleGraph.getModuleById("\0virtual:nav");
        if (!mod) return;
        s.reloadModule(mod);
      };
      s.watcher.add(["**/*.gsql", "**/*.md"]);
      s.watcher.on("all", refresh);
      refresh();
    }
  };
}
function updateParsedFiles(analysis) {
  workspaceFiles = workspaceFiles.map((file) => {
    let analyzed = analysis.files.find((next) => next.path == file.path);
    if (!analyzed) return file;
    return {
      ...file,
      parsed: {
        tree: analyzed.tree,
        virtualContents: analyzed.virtualContents,
        virtualToMarkdownOffset: analyzed.virtualToMarkdownOffset
      }
    };
  });
}
var handleRequestPlugin = {
  name: "handleRequest",
  configureServer: (s) => {
    s.middlewares.use(async function handleRequest(req, res, next) {
      try {
        let [pathName] = (req.url || "").split("?");
        if (pathName == "/_api/query") return await handleQuery(req, res);
        if (pathName) {
          if (pathName == "/__ct" || pathName == "/_charts" || pathName == "/_styles") return await handlePage(s, res);
        }
        if (!pathName || pathName == "/") pathName = "index";
        let relativeMdPath = pathName.replace(/^\//, "") + ".md";
        let mdPath = path2.join(config.root, relativeMdPath);
        if (mockFileMap[relativeMdPath] || await fs2.exists(mdPath)) {
          await handlePage(s, res);
        } else {
          next();
        }
      } catch (err) {
        if (process.env.NODE_ENV != "test") console.error(err);
        res.statusCode = 500;
        res.end(JSON.stringify({ message: err.message, stack: err.stack }));
      }
    });
  }
};
function mockFilesForTests() {
  if (process.env.NODE_ENV !== "test") return null;
  function toMockKey(id) {
    return id.replace(config.root + "/", "").replace(/^\//, "");
  }
  return {
    name: "mock-files-for-tests",
    enforce: "pre",
    resolveId(id) {
      if (!mockFileMap[toMockKey(id)]) return;
      return path2.join(config.root, toMockKey(id)) + "?mock";
    },
    load(id) {
      if (!id.endsWith("?mock")) return null;
      return mockFileMap[toMockKey(id.replace(/\?mock$/, ""))];
    }
  };
}
export {
  clearSvelteWarnings,
  prepareDeps,
  serve2,
  svelteWarnings
};
