// @graphenedata/cli/dist/cli/snowflake-HVSTYBLB.js
import {
  config
} from "./chunk-SQVXTHE5.js";

// connections/snowflake.ts
import { createPrivateKey } from "node:crypto";
import snowflake from "snowflake-sdk";
var SnowflakeConnection = class {
  ready;
  connection;
  constructor(opts) {
    this.ready = this.initialize(opts || {});
  }
  async initialize(opts) {
    let privateKeyPath = opts.privateKeyPath || config.snowflake?.privateKeyPath;
    let authOptions = {};
    if (privateKeyPath) {
      authOptions = { privateKeyPath, privateKeyPass: opts.privateKeyPass };
    } else if (opts.privateKey) {
      let privateKey = createPrivateKey({ key: opts.privateKey, format: "pem", passphrase: opts.privateKeyPass });
      authOptions = { privateKey: privateKey.export({ format: "pem", type: "pkcs8" }) };
    }
    snowflake.configure({ logLevel: opts.logLevel || "WARN", logFilePath: "/dev/null" });
    this.connection = snowflake.createConnection({
      ...opts,
      ...config.snowflake || {},
      ...authOptions,
      authenticator: "SNOWFLAKE_JWT",
      application: "Graphene"
    });
    await new Promise((resolve, reject) => {
      this.connection.connect((err, conn) => err ? reject(err) : resolve(conn));
    });
  }
  async runQuery(sql, params) {
    await this.ready;
    return await new Promise((resolve, reject) => {
      let rows = [];
      this.connection.execute({
        sqlText: sql,
        binds: params,
        streamResult: true,
        complete: (error, statement) => {
          if (error) {
            reject(new Error(`Snowflake query failed: ${error.message || error}`));
            return;
          }
          let stream = statement.streamRows();
          stream.on("error", (err) => reject(err));
          stream.on("readable", function(row) {
            while ((row = this.read()) !== null) {
              rows.push(row);
            }
          });
          stream.on("end", () => {
            let totalRows = Number(statement.getNumRows());
            resolve({ rows, totalRows });
          });
        }
      });
    });
  }
  async listDatasets() {
    let res = await this.runQuery("show databases");
    return res.rows.map((row) => String(row["name"] || ""));
  }
  async listSchemas(database) {
    let resolvedDatabase = await this.resolveDatabaseName(database);
    let res = await this.runQuery(`
      select schema_name as "schema_name"
      from ${snowflakeIdent(resolvedDatabase)}.INFORMATION_SCHEMA.SCHEMATA
      where schema_name != 'INFORMATION_SCHEMA'
      order by schema_name
    `);
    return res.rows.map((row) => String(row["schema_name"]).toLowerCase());
  }
  async listTables(dataset) {
    let parts = dataset.split(".");
    let database = await this.resolveDatabaseName(parts.shift() || "");
    let schema = parts.join(".");
    let res = await this.runQuery(
      `
      select table_schema as "table_schema", table_name as "table_name"
      from ${snowflakeIdent(database)}.INFORMATION_SCHEMA.TABLES
      where table_type in ('BASE TABLE', 'VIEW') and upper(table_schema) = upper(?)
      order by table_name
    `,
      [schema]
    );
    return res.rows.map((row) => `${String(row["table_schema"]).toLowerCase()}.${String(row["table_name"]).toLowerCase()}`);
  }
  async describeTable(target) {
    let parts = target.split(".");
    let database = await this.resolveDatabaseName(parts.shift() || "");
    let table = parts.pop() || "";
    let schema = parts.join(".");
    let res = await this.runQuery(
      `
      select column_name as "column_name", data_type as "data_type", ordinal_position as ordinal_position
      from ${snowflakeIdent(database)}.INFORMATION_SCHEMA.COLUMNS
      where upper(table_schema) = upper(?) and upper(table_name) = upper(?)
      order by ordinal_position
    `,
      [schema, table]
    );
    return res.rows.map((row) => {
      return { name: String(row["column_name"]).toLowerCase(), dataType: String(row["data_type"]) };
    });
  }
  async resolveDatabaseName(name) {
    let databases = await this.listDatasets();
    return databases.find((db) => db.toLowerCase() == name.toLowerCase()) || name;
  }
  async close() {
    await this.ready;
    await new Promise((resolve, reject) => {
      this.connection.destroy((err) => err ? reject(err) : resolve());
    });
  }
};
function snowflakeIdent(value) {
  if (!value) throw new Error("Snowflake identifiers cannot be empty");
  return `"${value.replace(/"/g, '""')}"`;
}
export {
  SnowflakeConnection
};
