// @graphenedata/html2canvas/dist/lib/core/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
var logger_1 = require("./logger");
var cache_storage_1 = require("./cache-storage");
var Context = /** @class */ (function () {
    function Context(options, windowBounds) {
        var _a, _b;
        this.windowBounds = windowBounds;
        this.instanceName = "#" + Context.instanceCount++;
        this.logger = new logger_1.Logger({ id: this.instanceName, enabled: options.logging });
        this.cache = (_a = options.cache) !== null && _a !== void 0 ? _a : new cache_storage_1.Cache(this, options);
        this.liveDOM = (_b = options.liveDOM) !== null && _b !== void 0 ? _b : false;
        this.cleanups = [];
    }
    Context.prototype.applyTemporaryStyle = function (element, property, value) {
        var propertyName = property;
        if (this.liveDOM) {
            var previous_1 = element.style.getPropertyValue(propertyName);
            var priority_1 = element.style.getPropertyPriority(propertyName);
            this.cleanups.push(function () {
                if (previous_1) {
                    element.style.setProperty(propertyName, previous_1, priority_1);
                }
                else {
                    element.style.removeProperty(propertyName);
                }
            });
        }
        element.style.setProperty(propertyName, value);
    };
    Context.prototype.cleanup = function () {
        while (this.cleanups.length) {
            var restore = this.cleanups.pop();
            if (restore) {
                restore();
            }
        }
    };
    Context.instanceCount = 1;
    return Context;
}());
exports.Context = Context;
